src/annotator/communicator.js

```diff
@@ -0,0 +1,375 @@
+ * - `guest/s`: iframes with annotatable content
+ *
+ * This layout represents the current arrangement of frames:
+ *
+ * `host` frame
+ * |-> (generally, shadow DOMed) `sidebar` iframe
+ * |-> (generally, shadow DOMed) `notebook` iframe
+ * |-> [`guest` iframe/s]
+ *     |-> [`guest` iframe/s]
+ *
+ * Currently, we support the communication between the following pair of frames:
+ * - `guest` <-> `sidebar` TODO
+ * - `host` <-> `sidebar` TODO
+ * - `notebook` <-> `sidebar`
+ *
+ * `FrameConnector` runs only on the `host` frame. The rest of the frames run the
+ * companion class, `PortFinder`. `FrameConnector` creates `MessageChannels` for
+ * two frames to communicate with each other. It also listens to requests for
+ * particular channel.port and dispatches the corresponding `MessagePort`.
+ *
+ *
+ *                 FrameConnector                      |                PortFinder
+ * ----------------------------------------------------|--------------------------------------------------------------
+ *
+ * 2. listens to `requests` of channel.port <--------------------- 1. `request` a channel.port using `host.postMessage`
+ *                 |
+ *                 V
+ * 3. sends `offers` of channel.port using `frame.postMessage` ---> 4. listens to `offers` of channel.port
+ *
+ * TODO: after `hostToSidebarChannel` has been established, it would be recommended
+ * in step #3 to substitute `frame.postMessage` for `sidebarPort.postMessage`
+ * using `hostToSidebarChannel.sidebarPort`.
+ *
+ * It is essential that `FrameConnect` initialize the listeners (step #2) before
+ * `PortFinder` sends the `requests` of channel.port (step #1). Because the
+ * `host` frame creates the `sidebar` and `notebook` iframes, it is guaranteed
+ * that `host` frame is ready to listen to messages originating from the
+ * `sidebar` and `notebook` iframes.
+ *
+ * However, the above assumption is not true for `host` and `guest` frames. Because
+ * `guest` iframes load in parallel, we can not assume that the code in the `host`
+ * frame is executed before the code in a `guest` frame. Therefore, for the `guest`
+ * frames we implement a polling strategy (sending a message every N milliseconds
+ * until a response is received).
+ *
+ *
+ * @implements Destroyable
+ */
+export class FrameConnector {
+  /**
+   * @param {object} options
+   *   @param {string} options.sidebarAppUrl
+   */
+  constructor({ sidebarAppUrl }) {
+    this._sidebarAndNotebookAppOrigin = new URL(sidebarAppUrl).origin;
+
+    // Create only the necessary channels. Channel nomenclature:
+    // `[frame1]To[frame2]Channel` so that `port1` should be owned by/sent to
+    // `frame1` and `port2` by `frame2`.
+    this._channels = {
+      /**
+       * Channels for the `guest` frame/s are created on demand (we don't know how
+       * many of those would be).
+       *
+       * @type {Array<{subFrameIdentifier: string, channel: MessageChannel}>}
+       */
+      guestToSidebarChannel: [], // TODO
```

Comment:
`subFrameIdentifier` could come handy to make sure that we don't create duplicated channels for the same `guest` frame. I believe the frame `origin` is not a replacement for `subFrameIdentifier`, because there could be two `guest` frames with the same origin. Maybe a reference to the `source` (window) could be used as the key on a map/dictionary<source, channel>.

For example:

1. `guest` frame starts polling the `host` to request its port.
2. `host` frame receives one of the request messages and creates a channel.
3. `host` sends the port to `guest`
4.  if `guest` sends another request to `host`, _after_ step 2, then it would create a duplicated channel.

Reply:
We can decide what is the best alternative to `subFrameIdentifier`.